src/lib/pubkey/ec_group/ec_inner_data.h

```diff
@@ -0,0 +1,247 @@
+      virtual ~EC_Scalar_Data() = default;
+
+      virtual const std::shared_ptr<const EC_Group_Data>& group() const = 0;
+
+      virtual size_t bytes() const = 0;
+
+      virtual std::unique_ptr<EC_Scalar_Data> clone() const = 0;
+
+      virtual bool is_zero() const = 0;
+
+      virtual bool is_eq(const EC_Scalar_Data& y) const = 0;
+
+      virtual void assign(const EC_Scalar_Data& y) = 0;
+
+      virtual void square_self() = 0;
+
+      virtual std::unique_ptr<EC_Scalar_Data> negate() const = 0;
+
+      virtual std::unique_ptr<EC_Scalar_Data> invert() const = 0;
+
+      virtual std::unique_ptr<EC_Scalar_Data> add(const EC_Scalar_Data& other) const = 0;
+
+      virtual std::unique_ptr<EC_Scalar_Data> sub(const EC_Scalar_Data& other) const = 0;
+
+      virtual std::unique_ptr<EC_Scalar_Data> mul(const EC_Scalar_Data& other) const = 0;
+
+      virtual void serialize_to(std::span<uint8_t> bytes) const = 0;
+};
+
+class EC_AffinePoint_Data {
+   public:
+      virtual ~EC_AffinePoint_Data() = default;
+
+      virtual const std::shared_ptr<const EC_Group_Data>& group() const = 0;
+
+      virtual std::unique_ptr<EC_AffinePoint_Data> clone() const = 0;
+
+      // Return size of a field element
+      virtual size_t field_element_bytes() const = 0;
+
+      // Writes 1 field element worth of data to bytes
+      virtual void serialize_x_to(std::span<uint8_t> bytes) const = 0;
+
+      // Writes 1 field element worth of data to bytes
+      virtual void serialize_y_to(std::span<uint8_t> bytes) const = 0;
+
+      // Writes 2 field elements worth of data to bytes
+      virtual void serialize_xy_to(std::span<uint8_t> bytes) const = 0;
+
+      // Writes 1 byte + 1 field element worth of data to bytes
+      virtual void serialize_compressed_to(std::span<uint8_t> bytes) const = 0;
+
+      // Writes 1 byte + 2 field elements worth of data to bytes
+      virtual void serialize_uncompressed_to(std::span<uint8_t> bytes) const = 0;
+
+      virtual std::unique_ptr<EC_AffinePoint_Data> mul(const EC_Scalar_Data& scalar,
+                                                       RandomNumberGenerator& rng,
+                                                       std::vector<BigInt>& ws) const = 0;
+
+      virtual EC_Point to_legacy_point() const = 0;
+};
+
+class EC_Mul2Table_Data {
+   public:
+      virtual ~EC_Mul2Table_Data() = default;
+
+      // Returns nullptr if g*x + h*y was point at infinity
+      virtual std::unique_ptr<EC_AffinePoint_Data> mul2_vartime(const EC_Scalar_Data& x,
+                                                                const EC_Scalar_Data& y) const = 0;
+
+      // Returns nullptr if g*x + h*y was point at infinity
+      virtual std::unique_ptr<EC_Scalar_Data> mul2_vartime_x_mod_order(const EC_Scalar_Data& x,
+                                                                       const EC_Scalar_Data& y) const = 0;
+};
+
+class EC_Group_Data final : public std::enable_shared_from_this<EC_Group_Data> {
```

Comment:
That would probably be cleaner. Main issue is that even with pcurves enabled we still have to support the full set of BigInt based APIs so we're still carrying around all of the same fields. I guess we could have the pcurves `EC_Group_Data` derive from the generic one to avoid too much duplication ...